File: backend/src/test/java/org/globe42/web/security/AdminOnlyIntegrationTest.java
package org.globe42.web.security;

import static org.mockito.Mockito.when;

import java.util.Optional;

import org.globe42.dao.UserDao;
import org.globe42.domain.User;
import org.globe42.web.exception.ForbiddenException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit4.SpringRunner;

/**
 * Integration test to verify that calling a method annotated with {@link AdminOnly} while not being an admin
 * indeed triggers the aspect and throws an exception
 * @author JB Nizet
 */
@SpringBootTest
@RunWith(SpringRunner.class)
@TestPropertySource("/test.properties")
public class AdminOnlyIntegrationTest {

    @MockBean
    private CurrentUser mockCurrentUser;

    @MockBean
    private UserDao mockUserDao;

    @SpyBean
    private AdminOnlyTester adminOnlyTester;

    @Before
    public void prepare() {
        long userId = 42L;
        when(mockCurrentUser.getUserId()).thenReturn(userId);
        User user = new User(userId);
        user.setAdmin(false);
        when(mockUserDao.findById(userId)).thenReturn(Optional.of(user));
    }

    @Test(expected = ForbiddenException.class)
    public void shouldThrowWhenCallingAdminOnlyAnnotatedMethod() {
        adminOnlyTester.foo();
    }

    @Test
    public void shouldNotThrowWhenCallingNotAnnotatedMethod() {
        adminOnlyTester.bar();
    }

    public static class AdminOnlyTester {
        @AdminOnly
        public void foo() {
        }

        public void bar() {
        }
    }
}

File: backend/src/test/kotlin/org/globe42/web/security/AdminOnlyIntegrationTest.kt
package org.globe42.web.security

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.globe42.dao.UserDao
import org.globe42.web.exception.ForbiddenException
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.EnableAspectJAutoProxy

/**
 * Integration test to verify that calling a method annotated with [AdminOnly] while not being an admin
 * indeed triggers the aspect and throws an exception
 * @author JB Nizet
 */
@SpringBootTest
class AdminOnlyIntegrationTest {

    @Autowired
    private lateinit var mockCurrentUser: CurrentUser

    @Autowired
    private lateinit var mockUserDao: UserDao

    @Autowired
    private lateinit var adminOnlyTester: AdminOnlyTester

    @BeforeEach
    fun prepare() {
        val userId = 42L
        every { mockCurrentUser.userId } returns userId
        every { mockUserDao.existsNotDeletedAdminById(userId) } returns false
    }

    @Test
    fun `should throw when calling admin only annotated method`() {
        assertThatExceptionOfType(ForbiddenException::class.java).isThrownBy { adminOnlyTester.foo() }
    }

    @Test
    fun `should not throw when calling not annotated method`() {
        adminOnlyTester.bar()
    }

    open class AdminOnlyTester {
        @AdminOnly
        open fun foo() {
        }

        open fun bar() {}
    }

    @Configuration
    @EnableAspectJAutoProxy
    class AlternateConfiguration {
        @Bean
        fun adminOnlyTester() = AdminOnlyTester()

        @Bean
        fun currentUser() = mockk<CurrentUser>()

        @Bean
        fun userDao() = mockk<UserDao>()

        @Bean
        fun adminOnlyAspect() = AdminOnlyAspect(currentUser(), userDao())
    }
}
